Do local write combining to avoid async http explosions

as-j opened the pull request:
Pack muliple event messages into a single influxdb write, and only
do a single async http call at once.  This also adds a stats page
to see what devices send huge number of events, and maybe help find
out why influxdb is a perceived source of many problems.

Issues:

1. runIn() will reschedule all subsequent calls, so this requires a 1s pause
between requests to actually complete a call, but for a 1st pass I'm ok with
this.

2. All messages are time stamped with local time before sending to influxdb
since calls could be delayed.  This simple and easy, so who not.  This should
probably use the event time as returned by evt.getUnixTime(), but this isn't
documented as to what it is, but this should be dug into since it surely
would be more accurate?

3. Influxdb line protocol (/write) is generated progressively as the line is
formed, and this is messy and could be centralized.  For example
postToInfluxDB() appends the timestamp to a completed line.  Also
postToInfluxDBNextFromQueue() specifies the precision which is implicit in the
earlier new Date().getTime() call.

4. There's potential synchronization issues with the queue if events are processed on
multiple threads.  Not

as-j committed:
Do local write combining to avoid async http explosions

Pack muliple event messages into a single influxdb write, and only
do a single async http call at once.  This also adds a stats page
to see what devices send huge number of events, and maybe help find
out why influxdb is a perceived source of many problems.

Issues:

1. runIn() will reschedule all subsequent calls, so this requires a 1s pause
between requests to actually complete a call, but for a 1st pass I'm ok with
this.

2. All messages are time stamped with local time before sending to influxdb
since calls could be delayed.  This simple and easy, so who not.  This should
probably use the event time as returned by evt.getUnixTime(), but this isn't
documented as to what it is, but this should be dug into since it surely
would be more accurate?

3. Influxdb line protocol (/write) is generated progressively as the line is
formed, and this is messy and could be centralized.  For example
postToInfluxDB() appends the timestamp to a completed line.  Also
postToInfluxDBNextFromQueue() specifies the precision which is implicit in the
earlier new Date().getTime() call.

ogiewon commented:
Hmmmm...  I just merged Caleb's Pull request...  His is definitely going to conflict with yours.  Perhaps the two of you could work together on making a common, best of both version?  

Really glad to see some developers trying to improve this App.  I have personally switched to using Node-Red + the Hubitat Events webSocket method of logging InfluxDB data.  So, your assistance is very much appreciated!

as-j committed:
Add static asyncLock to project threaded calls into queued data

This uses a static lock to protect the static data even if multiple
objects exist.

as-j committed:
Add send time stats and queue size stats

Plus make the stats dislpay easier to read

ogiewon commented:
Have you worked out any possible conflicts with Caleb’s PR?  I’ll happily merge your PR as long as you believe it won’t mess up any existing users who update the code.

dennypage commented:
This PR appears stale and has conflicts with existing code. If the PR is still pertinent, please resolve conflicts and re-open. Thanks.
